Marlin/Marlin_main.cpp

```diff
@@ -6018,130 +6022,161 @@ inline void gcode_M503() {
       SERIAL_ERRORLNPGM(MSG_TOO_COLD_FOR_M600);
       return;
     }
-
+  
+    // Show initial message and wait for synchronize steppers
+    lcd_filament_change_show_message(FILAMENT_CHANGE_MESSAGE_INIT);
+    st_synchronize();
+  
     float lastpos[NUM_AXIS];
-    #if ENABLED(DELTA)
-      float fr60 = feedrate / 60;
-    #endif
-
+  
+    // Save current position of all axes
     for (int i = 0; i < NUM_AXIS; i++)
       lastpos[i] = destination[i] = current_position[i];
-
+  
+    // Define runplan for move axes
     #if ENABLED(DELTA)
       #define RUNPLAN calculate_delta(destination); \
-                      plan_buffer_line(delta[X_AXIS], delta[Y_AXIS], delta[Z_AXIS], destination[E_AXIS], fr60, active_extruder);
+        plan_buffer_line(delta[X_AXIS], delta[Y_AXIS], delta[Z_AXIS], destination[E_AXIS], FILAMENT_CHANGE_XY_FEEDRATE * 60, active_extruder);
     #else
-      #define RUNPLAN line_to_destination();
+      #define RUNPLAN line_to_destination(FILAMENT_CHANGE_XY_FEEDRATE * 60);
     #endif
-
-    //retract by E
+  
+    KEEPALIVE_STATE(IN_HANDLER);
+  
+    // Initial retract before move to filament change position
     if (code_seen('E')) destination[E_AXIS] += code_value();
-    #ifdef FILAMENTCHANGE_FIRSTRETRACT
-      else destination[E_AXIS] += FILAMENTCHANGE_FIRSTRETRACT;
-    #endif
-
-    RUNPLAN;
-
-    //lift Z
+      #ifdef FILAMENT_CHANGE_RETRACT_LENGTH
+        else destination[E_AXIS] -= FILAMENT_CHANGE_RETRACT_LENGTH;
+      #endif
+    line_to_destination(FILAMENT_CHANGE_RETRACT_FEEDRATE * 60);
+  
+    // Lift Z axis
     if (code_seen('Z')) destination[Z_AXIS] += code_value();
-    #ifdef FILAMENTCHANGE_ZADD
-      else destination[Z_AXIS] += FILAMENTCHANGE_ZADD;
+      #ifdef FILAMENT_CHANGE_Z_ADD
+        else destination[Z_AXIS] += FILAMENT_CHANGE_Z_ADD;
```

Comment:
An improvement suggestion, this should check if the move will not go over Z-Max.